profiling/src/lib.rs

```diff
@@ -1101,7 +1121,17 @@ unsafe extern "C" fn alloc_profiling_realloc(
 ) -> *mut ::libc::c_void {
     let ptr: *mut libc::c_void;
     if PREV_CUSTOM_MM_REALLOC.is_none() {
-        ptr = zend::_zend_mm_realloc(zend::zend_mm_get_heap(), prev_ptr, len);
+        // the first element of the heap struct is the `use_custom_heap` flag. We got to reset this
+        // to 0 to allow for `zend_mm_gc()` to do garbage collection when called during allocation
+        // TODO: find out the correct size of that `use_custom_heap` element at this point I am
+        // assuming that it is a 32 bit integer ....
+        let heap = zend::zend_mm_get_heap();
+        let custom_heap: u32 = std::ptr::read(heap as *const u32);
+        std::ptr::write(heap as *mut u32, zend::ZEND_MM_CUSTOM_HEAP_NONE);
```

Comment:
This highlighted code seems to be repeated quite a few times. Can we extract a function for it?

Reply:
I refactored the preparation and restore into their own functions. Also added a nice comment to make clear what we are doing, why we are doing it, and why we may consider it okay (I don't want to use the word safe in this case 😉 )